src/fastpath/fastpath.c

```diff
@@ -6,15 +6,269 @@
+            if (!ntfn_queue.head) {
+                notification_ptr_set_state(ntfnPtr, NtfnState_Idle);
+            }
+
+            maybeDonateSchedContext(dest, ntfnPtr);
+
+            /* Get destination thread VTable */
+            newVTable = TCB_PTR_CTE_PTR(dest, tcbVTable)->cap;
+
+            /* Ensure that the destination has a valid VTable. */
+            if (unlikely(! isValidVTableRoot_fp(newVTable))) {
+                slowpath(SysSend);
+            }
+
+            /* Get vspace root. */
+            cap_pd = cap_vtable_cap_get_vspace_root_fp(newVTable);
+
+#ifdef CONFIG_ARCH_AARCH32
+            /* Get HW ASID */
+            stored_hw_asid = cap_pd[PD_ASID_SLOT];
+#endif
+
+#ifdef CONFIG_ARCH_X86_64
+                    /* borrow the stored_hw_asid for PCID */
+                    stored_hw_asid.words[0] = cap_pml4_cap_get_capPML4MappedASID_fp(newVTable);
+#endif
+
+#ifdef CONFIG_ARCH_IA32
+                    /* stored_hw_asid is unused on ia32 fastpath, but gets passed into a function below. */
+                    stored_hw_asid.words[0] = 0;
+#endif
+#ifdef CONFIG_ARCH_AARCH64
+                    stored_hw_asid.words[0] = cap_vtable_root_get_mappedASID(newVTable);
+#endif
+
+#ifdef CONFIG_ARCH_RISCV
+                    /* Get HW ASID */
+                    stored_hw_asid.words[0] = cap_page_table_cap_get_capPTMappedASID(newVTable);
+#endif
+
+            /* Let gcc optimise this out for 1 domain */
+            dom = maxDom ? ksCurDomain : 0;
+            /* Ensure only the idle thread or lower prio threads are present in the scheduler */
+            if (unlikely(dest->tcbPriority < NODE_STATE(ksCurThread->tcbPriority) &&
+                         !isHighestPrio(dom, dest->tcbPriority))) {
+                slowpath(SysSend);
+            }
+
+#ifdef CONFIG_ARCH_AARCH32
+            if (unlikely(!pde_pde_invalid_get_stored_asid_valid(stored_hw_asid))) {
+                slowpath(SysSend);
+            }
+#endif
+
+            /* Ensure the destination thread is in the current domain and can be scheduled directly. */
+            if (unlikely(dest->tcbDomain != ksCurDomain && maxDom)) {
+                slowpath(SysSend);
+            }
+
+#ifdef ENABLE_SMP_SUPPORT
+            /* Ensure both threads have the same affinity */
+            if (unlikely(NODE_STATE(ksCurThread)->tcbAffinity != dest->tcbAffinity)) {
+                slowpath(SysSend);
+            }
+#endif /* ENABLE_SMP_SUPPORT */
+
+#ifdef CONFIG_BENCHMARK_TRACK_KERNEL_ENTRIES
+            ksKernelEntry.is_fastpath = true;
+#endif
+
+        thread_state_ptr_set_tsType_np(&dest->tcbState, ThreadState_Running);
+        setRegister(dest, badgeRegister, badge);
+        if (ksCurThread->tcbPriority < dest->tcbPriority) {
+            /* switch to waiter immediately */
+            SCHED_ENQUEUE_CURRENT_TCB;
+            switchToThread_fp(dest, cap_pd, stored_hw_asid);
+            NODE_STATE(ksCurSC) = NODE_STATE(ksCurThread)->tcbSchedContext;
```

Comment:
where does the scheduling context get charged?

Reply:
I'm not sure if/where it does currently. From what I understood this SC donation is similar to [how we do it in `fastpath_call`](https://github.com/seL4/seL4/blob/master/src/fastpath/fastpath.c#L184-L187), isn't it? Have I misunderstood something?

Reply:
Hm, might this happen in the call to [`maybeDonateSchedulingContext`](https://github.com/seL4/seL4/pull/322/files#diff-931125ed3a85619bdb93c0e1a4f69f62af97b5c815d986d0270d1cfb5a5209e5R188).

Reply:
When we donate a scheduling context, no charging happens, as we are not changing the scheduling context.

however, if we change scheduling context, we need to charge the previous one for the time that has passed (or we'll lose it). 

So far we don't fastpath anything that doesn't donate a scheduling context. I am not sure what the intent of this fastpath is. It needs to be some cut down version of the slowpath, likely the most simple form of charging a scheduling context (I *think* this is what I did in the irq fastpath on the branch I linked, but I don't have time to check now). 

Do you have a high level statement of what this fastpath should be fastpathing? Ideally its a small subset of things that happen on the slowpath.

Reply:
My above statement is wrong. Even if we do `maybeDonateSchedulingContext` this donates from the notification object to the thread - we're still changing the current SC. So we have to charge.

My recommendation is you find the fastest path through switching scheduling context and fastpath that and only that.

Reply:
> Do you have a high level statement of what this fastpath should be fastpathing? Ideally its a small subset of things that happen on the slowpath.

At a high level, the intention of this fastpath is to fastpath signalling a lower priority thread, and signalling a higher (or equal) priority thread in cases where the signaller and signalled thread are in the same scheduling domain. You may find [the thesis report](https://drive.google.com/file/d/143Quc9KlXeYT6p6k2ask22zCkj-c9VTv/view?usp=sharing) helpful for a bit more context (in particular the flowchart at Figure 4.4, though now I realise that the 'Donate scheduling context from signaller to waiter' stage should rather say 'Donate scheduling context from _notification_ to waiter').

For my understanding; are you saying that in cases where we switch to the signalled thread, we need to charge the signaller's SC for the time elapsed? Based off your comment and a quick dig through the code it looks like [`SCHED_ENQUEUE_CURRENT_TCB`](https://github.com/shanekadish/seL4/blob/signal-fastpath/src/fastpath/fastpath.c#L258) does _not_ handle this charging logic, right? So we need to explicitly charge the signaller's SC before/after invoking that?

It's not clear to me where this charging is happening in the signal slowpath. Is it done in something downstream of [`possibleSwitchTo(dest)`](https://github.com/shanekadish/seL4/blob/signal-fastpath/src/object/notification.c#L116-L118)?